### DISCUSSION TITLE
Decoding an apparently fine QRCode leads to "failed to read version" exception.

### DISCUSSION THREAD

**Discussion by rderimay** [chillerlan/php-qrcode#271] [Feedback]
**Describe the bug or unexpected behaviour**

A QRCode recognition fails with the error "failed to read version" even if the QRCode can be decoded with a phone for example. 

**Steps to reproduce the behavior**

 `(new QRCode)->readFromFile($imagePath);`

with `imagePath` being a good resolution jpg file.

It fails in file `src/Decoder/BitMatrix.php`, line 345

If you provide me with a way to send you the file privately  or a way to track down the problem I would be happy to help. The QRCode contains personal information of someone else, so I cannot post the file in public here.

**Environment (please complete the following information):**

- PHP version/OS: PHP 8.3, MacOS or Debian buster, makes no difference.
- Library version: 5.0.2

**Comment by codemasher** [comment]
Hi, first off: without the image producing that error I'm not in a position to help in any way. 
Second, the decoder does not work in the same way as it would on a mobile app - the mobile app scans the code continuously from different angles and positions and will return the first positive result it gets, while the decoder here only does a single pass on the fixed source image.
The decoding algorithm is very similar to the mobile apps and in theory it could be implemented in a way to multi pass scan an image while transforming the source image, but that's not the focus of this library.

**Comment by rderimay** [comment]
Yes off course. Is there a way I could send it to you in a private way ?
Also, I could crop and resize the image multiple times before sending it the lib. Would that help ? if yes, what would be the best transformations (change size ? change offset ? change rotation ?) to apply to maximise the chance to have a positive ?
Thanks !

**Comment by codemasher** [comment]
> Is there a way I could send it to you in a private way ?

My email address is in the file header of almost every file in this library.


> Also, I could crop and resize the image multiple times before sending it the lib. Would that help ?

I assume you mean before decoding? There's many things that you could do, first, the grayscale and contrast options, rotation, resizing ([smaller QR Codes produce better results for some reason](https://github.com/chillerlan/php-qrcode/blob/0ca3d2377f8d01b788c738d46f9bf72d9b17083d/tests/QRCodeReaderTestAbstract.php#L156-L160)), [perspective transform](https://github.com/chillerlan/php-qrcode/blob/main/src/Detector/PerspectiveTransform.php) (tilt along the axes).

Basically you could emulate the scanning of a mobile camera somewhat like this:


```php
while(true){
    try{
        // read qrcode ...
        return $result;
    }
    catch(Throwable){
        // apply transforms to image, continue, break after X retries maybe
    }
}
```

**Comment by codemasher** [comment]
> in theory it could be implemented in a way to multi pass scan an image while transforming the source image

I should add here that it's hard to estimate the correct directions for a perspective transform. Unlike a human, who looks at the output of the mobile camera's capturing input and who can position the QR Code in the app's view finder by tilting the camera in 3 axes, an algorithm here could only guess and try tilting the image back and forth a bit in hope to get a result.

**Reply by rderimay** [reply]
Given your input I have made numerous tests changing the size of the image, changing it to monochrome or not, rotating it more or less. None of them has been leading to a positive detection.
I have sent the file to your email address. Maybe there is something particular that makes it difficult to be recognised and that I could improve before sending it to the lib...
Thanks for you expertise...

**Reply by rderimay** [reply]
@codemasher Sorry, I just found a bug in my script leading to the same image being sent to the lib instead of the differently modified ones.
Don't loose your time, I will make further tests myself to see if something work...

**Reply by codemasher** [reply]
Heh, ok, I'll check it out by myself too when I got some time on my hands. Good luck!

**Reply by codemasher** [reply]
Ok, I got a result: I cut out the QR Code of the source image with a margin of ~4-5 pixels (quiet zone), upscaled it (200%) and sharpened the resulting output, which gave me the URL contained in the QR Code.

**Reply by codemasher** [reply]
So just to confirm the reader capabilities, i tried this:

```php
$options = new QROptions;
$options->readerUseImagickIfAvailable = true;
$options->readerGrayscale             = false;
$options->readerIncreaseContrast      = false;

$options->outputInterface = QRGdImagePNG::class;
$options->outputBase64    = false;
$options->scale           = 2;

$qrcode = new QRCode($options);

try{
	// read the original source image
	$result = $qrcode->readFromFile(__DIR__.'/test-2000-a.png');
	// render qrcode from the result matrix
	$img = $qrcode->renderMatrix($result->getQRMatrix(), __DIR__.'/test-2000-b.png');
	// read from newly rendered
	$result = $qrcode->readFromBlob($img);

	var_dump($result);
}
catch(Throwable $e){
	echo $e->getMessage();
}
```

The newly generated QR code is 2 pixels per module sharp and read perfectly fine even without quiet zone. It seems that the source image you sent me was larger in the original and has been resized, resulting in an unsharp QR Code as well as the rest of the content (the ZXing mobile app also took unusually long to detect it), so I think that may be the issue. A recommendation would be to render the QR Code with ECC level L and a smaller version, which results in a smaller QR symbol, which in turn can be rendered with larger modules (3-5 pixel per module) to increase general readability.

**Reply by rderimay** [reply]
Yes. Somehow, the QR code in the original pdf is no vector graphic, but a somewhat unsharp image. Making a loop a tries with different sizes and rotations has lead to the library finding the qr code :-)
Many thanks for your help !

**Comment by codemasher** [comment]
Glad to be of help!

Here's the resizing i tired in Imagemagick (for anyone else with a similar issue):

```php
$im = new Imagick(__DIR__.'/test-2000-a.png');

$im->resizeImage(
	($im->getImageWidth() * 2),
	($im->getImageHeight() * 2),
	Imagick::FILTER_ROBIDOUXSHARP,
	1.0,
);

$im->sharpenImage(3, 0.5);

// save intermediate image to check modifications
$im->writeImage(__DIR__.'/im_test.png');

$result = (new Decoder($options))->decode(new IMagickLuminanceSource($im, $options));
```
